# Homework 3

## Recitation Exercises

### 1.1 Chapter 13

<br>

#### Exercise 1: Given the following points: 2, 4, 10, 12, 3, 20, 30, 11, 25. Assume $k=3$ and that we randomly pick the initial means $\mu_1=2$, $\mu_2=4$, and $\mu_3=6$. Show the clusters obtained using K-means algorithm after one iteration, and show the new means for the next iteration. 

<br>

Initial cluster:
<br>
$C_1 = {2, 3}$
<br>
$C_2 = {4}$
<br>
$C_3 = {10, 11, 12, , 20, 25, 30}$

<br>

Mean of the new clusters:
<br>
$\mu_1 = \frac{2+3}{2} = 2.5$
<br>
$\mu_2 = \frac{4}{1} = 4$
<br>
$\mu_3 = \frac{10+11+12+20+25+30}{6} = \frac{108}{6} = 18$

<br>

The new means for the next iteration are: $\mu_1 = 2.5$, $\mu_2 = 4$, and $\mu_3 = 18$

<br>

#### Exercise 3: Given the two-dimensional points in Table 13.2, assume that $k=2$, and that initially the points are assigned to clusters as follows: $C_1={x_1, x_2, x_4}$ and $C_2={x_3, x_5}$. Answer the following questions:

<br>

$$
\begin{array}{|c|c|c|}
\hline
& X_1 & X_2\\
\hline
x^{T}_{1} & 0 & 2\\
x^{T}_{2} & 0 & 0\\
x^{T}_{3} & 1.5 & 0\\
x^{T}_{4} & 5 & 0\\
x^{T}_{5} & 5 & 2 \\
\hline
\end{array}
$$

<br>

#### a.) Apply the K-means algorithm until convergence, that is, the clusters do not change, assuming (1) the usual Euclidian distance or the $L_2$-norm as the distance between points, defined as $||x_i - x_j||_{2}=(\sum^{d}_{a=1} (x_{ia} - x_{ja})^2)^{\frac{1}{2}}$, and (2) the Manhattan distance or the $L_1$-norm defined as $||x_i - x_j||_{1}=\sum^{d}_{a=1} |x_{ia} - x_{ja}|$.

<br>

The given data points are: $x_1(0,2)$, $x_2(0,0)$, $x_3(1.5,0)$, $x_4(5,0)$, $x_5(5,2)$ and the two clusters are $C_1(x_1, x_2, x_4)$ and $C_2(x_3, x_5)$. 
<br>
Initialization of the means: $\mu_1=(1.66, 0.66)$ and $\mu_2=(3.25, 1)$

<br>

Suppose $D_{ij}$ represents the distance between the mean of the ith and jth data point. Using the Euclidian distance, the distance between points and means of the clusters calculated are as below:
<br>
Distance of $x_1$ from the mean of clusters is: $D_{11}=2.13$, $D_{21}=3.40$
<br>
Distance of $x_2$ from the mean of clusters is: $D_{12}=1.78$, $D_{22}=3.40$
<br>
Distance of $x_3$ from the mean of clusters is: $D_{13}=3.40$, $D_{23}=2.01$
<br>
Distance of $x_4$ from the mean of clusters is: $D_{14}=0.67$, $D_{24}=2.01$
<br>
Distance of $x_5$ from the mean of clusters is: $D_{15}=3.59$, $D_{25}=2.01$
<br>

Looking at the above distances, we can see that points $x_1$, $x_2$, and $x_4$ have less distance to cluster 1 and data points $x_3$ and $x_5$ have less distance to cluster 2. There is no change in the clusters therefore convergence is achieved. 

<br> 

Below are the distance between points and means of the clusters using the Manhattan distance:
<br>
Distance of $x_1$ from the mean of clusters is: $D_{11}=3$, $D_{21}=4.25$
<br>
Distance of $x_2$ from the mean of clusters is: $D_{12}=2.32$, $D_{22}=4.25$
<br>
Distance of $x_3$ from the mean of clusters is: $D_{13}=4$, $D_{23}=2.75$
<br>
Distance of $x_4$ from the mean of clusters is: $D_{14}=0.82$, $D_{24}=2.75$
<br>
Distance of $x_5$ from the mean of clusters is: $D_{15}=4.68$, $D_{25}=2.75$
<br>

Looking at the above distances, we can see that points $x_1$, $x_2$, and $x_4$ have less distance to cluster 1 and data points $x_3$ and $x_5$ have less distance to cluster 2. There is no change in the clusters therefore convergence is achieved. 

<br>

#### b.) Apply the EM algorithm with $k=2$ assuming that the dimensions are independent. Show one complete execution of the expectation and maximization steps. Start with the assumption that $P(C_i | x_{ja})=0.5$ for $a=1,2$ and $j=1,...,5$.

<br>

The given data points are: $x_1(0,2)$, $x_2(0,0)$, $x_3(1.5,0)$, $x_4(5,0)$, $x_5(5,2)$ and the two clusters are $C_1(x_1, x_2, x_4)$ and $C_2(x_3, x_5)$. 
<br>
Initialization of the mean, covariance matrix, and probability:
<br> 
$\mu_1=(1.66, 0.66)$ and $\mu_2=(3.25, 1)$
<br>
$
Cov_1 = 
\begin{pmatrix}
1 & 0\\
0 & 1
\end{pmatrix}
$ 
and 
$Cov_2 = 
\begin{pmatrix}
1 & 0\\
0 & 1
\end{pmatrix}
$
<br>
$P(C_1)=0.5$ and $P(C_2)=0.5$

<br>

In the expectation step, the posterior probability of each point for each cluster is calculated. The probability of $x_j$ belonging to the cluster $C_i$ is given:
$$P(C_i|x_j)=\frac{f_i (x_j | \mu_i, \sigma_i^2) P(C_i)}{\sum^{k}_{a=1} f(x_j | \mu_a, \sigma_a^2) \cdot P(C_a)}$$ 
where $f(x|\mu_i, \sigma^2_i) = \frac{1}{\sqrt{2\pi}\sigma_i}e^{-\frac{(x-\mu_i)^2}{2\sigma^2_i}}$
<br>

The univariate normal $x_1$ and both clusters: 
<br>
$f(x_1 |\mu_1, \sigma^2_1)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_1 - \mu_1||}{2}}=0.13$
<br>
$f(x_1 |\mu_2, \sigma^2_2)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_1 - \mu_2||}{2}}=0.07$

<br>

The univariate normal $x_2$ and both clusters: 
<br>
$f(x_2 |\mu_1, \sigma^2_1)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_2 - \mu_1||}{2}}=0.16$
<br>
$f(x_2 |\mu_2, \sigma^2_2)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_2 - \mu_2||}{2}}=0.07$

<br>

The univariate normal $x_3$ and both clusters: 
<br>
$f(x_3 |\mu_1, \sigma^2_1)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_3 - \mu_1||}{2}}=0.07$
<br>
$f(x_3 |\mu_2, \sigma^2_2)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_3 - \mu_2||}{2}}=0.14$

<br>

The univariate normal $x_4$ and both clusters: 
<br>
$f(x_4 |\mu_1, \sigma^2_1)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_4 - \mu_1||}{2}}=0.28$
<br>
$f(x_4 |\mu_2, \sigma^2_2)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_4 - \mu_2||}{2}}=0.14$

<br>

The univariate normal $x_5$ and both clusters:
<br> 
$f(x_5 |\mu_1, \sigma^2_1)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_5 - \mu_1||}{2}}=0.06$
<br>
$f(x_5 |\mu_2, \sigma^2_2)=\frac{1}{\sqrt{2\pi}} e^{-\frac{||x_5 - \mu_2||}{2}}=0.14$

<br>

The probability of $x_1$ belonging to both can be calculated as: 
<br>
$P(C_1 | x_1)=\frac{0.13 \times 0.5}{0.13 \times 0.5 + 0.16 \times 0.5 + 0.07 \times 0.5 + 0.28 \times 0.5 + 0.06 \times 0.5}=0.18$
<br>
$P(C_2 | x_1)=\frac{0.07 \times 0.5}{0.07 \times 0.5 + 0.07 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5}=0.12$

<br>

The probability of $x_2$ belonging to both can be calculated as: 
<br>
$P(C_1 | x_2)=\frac{0.16 \times 0.5}{0.13 \times 0.5 + 0.16 \times 0.5 + 0.07 \times 0.5 + 0.28 \times 0.5 + 0.06 \times 0.5}=0.22$
<br>
$P(C_2 | x_2)=\frac{0.07 \times 0.5}{0.07 \times 0.5 + 0.07 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5}=0.12$

<br>

The probability of $x_3$ belonging to both can be calculated as: 
<br>
$P(C_1 | x_3)=\frac{0.07 \times 0.5}{0.13 \times 0.5 + 0.16 \times 0.5 + 0.07 \times 0.5 + 0.28 \times 0.5 + 0.06 \times 0.5}=0.1$
<br>
$P(C_2 | x_3)=\frac{0.14 \times 0.5}{0.07 \times 0.5 + 0.07 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5}=0.25$

<br>

The probability of $x_4$ belonging to both can be calculated as: 
<br>
$P(C_1 | x_4)=\frac{0.28 \times 0.5}{0.13 \times 0.5 + 0.16 \times 0.5 + 0.07 \times 0.5 + 0.28 \times 0.5 + 0.06 \times 0.5}=0.4$
<br>
$P(C_2 | x_4)=\frac{0.14 \times 0.5}{0.07 \times 0.5 + 0.07 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5}=0.25$

<br>

The probability of $x_5$ belonging to both can be calculated as: 
<br>
$P(C_1 | x_5)=\frac{0.06 \times 0.5}{0.13 \times 0.5 + 0.16 \times 0.5 + 0.07 \times 0.5 + 0.28 \times 0.5 + 0.06 \times 0.5}=0.08$
<br>
$P(C_2 | x_5)=\frac{0.14 \times 0.5}{0.07 \times 0.5 + 0.07 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5 + 0.14 \times 0.5}=0.25$

<br>

Therefore, the weights for $C_1$ are $w_1 = (0.18, 0.22, 0.1, 0.4, 0.08)$ and the weights for $C_2$ are $w_2 = (0.12, 0.12, 0.25, 0.25, 0.25)$

<br>

In the maximization step, mean and probability are re-estimated.
<br>
The mean is calculated as:
$$\mu_i = \frac{\sum^{n}_{j=1} w_{ij} \cdot x_j}{\sum^{n}_{j=1} w_{ij}}$$
therefore, $\mu_1 = 3.13$ and $\mu_2 = 3.64$

<br>

The probability can be re-estimated as:
$$P(C_i)=\frac{\sum^{n}_{j=1}w_{ij}}{n}$$
therefore, $P(C_1)=0.196$ and $P(C_2)=0.198$

### 1.2 Chapter 14

<br>

#### Exercise 1: Consider the 5-dimensional categorical data shown in Table 14.3 
$$
\begin{array}{|c|c|c|c|c|c|}
\hline
\text{Point} & X_1 & X_2 & X_3 & X_4 & X_5\\
\hline
x^{T}_{1} & 1 & 0 & 1 & 1 & 0\\
x^{T}_{2} & 1 & 1 & 0 & 1 & 0\\
x^{T}_{3} & 0 & 0 & 1 & 1 & 0\\
x^{T}_{4} & 0 & 1 & 0 & 1 & 0\\
x^{T}_{5} & 1 & 0 & 1 & 0 & 1 \\
x^{T}_{6} & 0 & 1 & 1 & 0 & 0 \\
\hline
\end{array}
$$

#### The similarity between categorical data points can be computed in terms of the number of matches and mismatches for the different attributes. Let $n_{11}$ be the number of attributes on which two points $x_i$ and $x_j$ assume the value 1, and let $n_{10}$ denote the number of attributes where $x_i$ takes value 1, but $x_j$ takes on the value of 0. Define $n_{01}$ and $n_{00}$ in a similar manner. The contingency table for measuring the similarity is then given as
$$
\begin{array}{|c|c|c|c|}
\hline
& x_j & x_j & x_j \\
\hline
x_i & & 1 & 0 \\
\hline
x_i & 1 & n_{11} & n_{10} \\
\hline
x_i & 0 & n_{01} & n_{00} \\
\hline
\end{array}
$$
#### Define the following similarity measures
* #### Simple matching coefficient: $SMC(x_i, x_j)=\frac{n_{11}+n_{00}}{n_{11}+n_{10}+n_{01}+n_{00}}$
* #### Jaccard coefficient: $JC(x_i, x_j)=\frac{n_{11}}{n_{11}+n_{10}+n_{01}}$
* #### Rao's coefficient: $RC(x_i, x_j)=\frac{n_{11}}{n_{11}+n_{10}+n_{01}+n_{00}}$

<br>

#### Find the cluster dendrograms produced by the hierarchical clustering algorithm under the following scenarios:
<br>

#### a.) We use single link with RC 
<br>

$$
\begin{array}{|c|c|c|c|c|c|}
\hline
 & X_2 & X_3 & X_4 & X_5 & X_6 \\
X_1 & 3/5 & 3/5 & 4/5 & 3/5 & 4/5\\
X_2 & & 4/5 & 3/5 & 4/5 & 4/5\\
X_3 & & & 4/5 & 4/5 & 4/5\\
X_4 & & & & 5/5 & 4/5\\
X_5 & & & & & 4/5\\
\hline
\end{array}
$$


$$
\begin{array}{|c|c|c|c|c|c|}
\hline
& & X_3 & X_4 & X_5 & X_6\\
X_1, X_2 & & 3/5 & 3/5 & 3/5 & 4/5\\
X_3 & & & 4/5 & 4/5 & 4/5\\
X_4 & & & & 5/5 & 4/5\\
X_5 & & & & & 4/5\\
\hline
\end{array}
$$

$$
\begin{array}{|c|c|c|c|c|c|}
\hline
 & & & X_4 & X_5 & X_6\\
X_1, X_2, X_3 & & & 3/5 & 3/5 & 4/5\\
X_4 & & & & 5/5 & 4/5\\
X_5 & & & & & 4/5\\
\hline
\end{array}
$$

the dendrogram is created below:
<br>
```commandline
1   2   3   4   5   6       Distance:
|   |   |   |   |   |           
 12     |   |   |   |           3/5
  |     |   |   |   |
    123     |   |   |           3/5
     |      |   |   |
       1234     |   |           3/5
         |      |   |       
           12345    |           3/5
             |      |
              123456            4/5
```
<br>

#### b.) We use complete link with SMC
<br>

$$
\begin{array}{|c|c|c|c|c|c|}
\hline
 & X_2 & X_3 & X_4 & X_5 & X_6 \\
X_1 & 2/5 & 1/5 & 3/5 & 2/5 & 3/5\\
X_2 & & 3/5 & 1/5 & 4/5 & 3/5\\
X_3 & & & 2/5 & 3/5 & 2/5\\
X_4 & & & & 5/5 & 2/5\\
X_5 & & & & & 3/5\\
\hline
\end{array}
$$


$$
\begin{array}{|c|c|c|c|c|c|}
\hline
& & & X_2, X_4 & X_5 & X_6\\
X_1, X_3 & & & 3/5 & 3/5 & 3/5\\
X_2, X_4 & & & & 5/5 & 3/5\\
X_5 & & & & & 3/5\\
\hline
\end{array}
$$

$$
\begin{array}{|c|c|c|c|c|c|}
\hline
 & & & & X_5 & X_6\\
X_1, X_2, X_3, X_4 & & & & 5/5 & 3/5\\
X_5 & & & & & 3/5\\
\hline
\end{array}
$$

the dendrogram is created below:
<br>
```commandline
1   3   2   4   6   5       Distance:
|   |   |   |   |   |
  13      24    |   |           1/5
   |       |    |   |
      1234      |   |           3/5
        |       |   |
          12346     |           3/5
            |       |
             123456             5/5
```



#### c.) We use group average with JC
<br>

$$
\begin{array}{|c|c|c|c|c|c|}
\hline
 & X_2 & X_3 & X_4 & X_5 & X_6 \\
X_1 & 2/4 & 1/3 & 3/4 & 2/4 & 3/4\\
X_2 & & 3/4 & 1/3 & 4/5 & 3/4\\
X_3 & & & 2/3 & 3/4 & 2/3\\
X_4 & & & & 5/5 & 2/3\\
X_5 & & & & & 3/4\\
\hline
\end{array}
$$


$$
\begin{array}{|c|c|c|c|c|c|}
\hline
& & & X_2, X_4 & X_5 & X_6\\
X_1, X_3 & & & 0.67 & 0.625 & 0.708\\
X_2, X_4 & & & & 0.9 & 0.708\\
X_5 & & & & & 0.6\\
\hline
\end{array}
$$

$$
\begin{array}{|c|c|c|c|c|c|}
\hline
 & & & & X_2, X_4 & X_5, X_6\\
X_1, X_3 & & & & 0.625 & 0.666\\
X_2, X_4 & & & & & 0.804\\
\hline
\end{array}
$$

the dendrogram is created below:
```commandline
1   3   2   4   5   6       Distance:
|   |   |   |   |   |
 13       24    |   |           1/3
  |        |    |   |
  |        |      56            0.6
  |        |       |
     1234          |            0.625
       |           |
           123456
```


#### Exercise 3: Using the distance matrix from Table 14.4, use the average link method to generate hierarchical clusters. Show the merging distance thresholds. 

$$
\begin{array}{|c|c|c|c|c|c|}
\hline
& A & B & C & D & E\\
\hline
A & 0 & 1 & 3 & 2 & 4 \\
\hline
B & & 0 & 3 & 2 & 3 \\
\hline
C & & & 0 & 1 & 3 \\
\hline
D & & & & 0 & 5 \\
\hline
E & & & & & 0 \\
\hline
\end{array}
$$

<br>

The minimum distance from the given dataset is 1 at $A,B$. Now we generate the distance matrix for the $AB$ cluster:
<br>

The distance from $AB$ to $C$,
$$
distance(AB, C)=average(distance(AB, C))=\frac{distance(A, C) + distance(B, C)}{2}=\frac{3+3}{2}=3
$$

<br>

The distance from $AB$ to $D$,
$$
distance(AB, D)=average(distance(AB, D))=\frac{distance(A, D) + distance(B, D)}{2}=\frac{2+2}{2}=2
$$

<br>

The distance from $AB$ to $E$,
$$
distance(AB, E)=average(distance(AB, E))=\frac{distance(A, E) + distance(B, E)}{2}=\frac{4+3}{2}=3.5
$$

<br>

the new distance matrix with cluster $AB$:

\begin{array}{|c|c|c|c|c|}
\hline
& AB & C & D & E\\
\hline
AB & 0 & 3 & 2 & 3.5\\
\hline
C & & 0 & 1 & 3\\
\hline
D & & & 0 & 5\\
\hline
E & & & & 0\\
\hline
\end{array}

<br>


Now we find the next minimum distance from table 2 to form the second cluster. The minimum distance from the table is 1 at $C,D$. Now we generate the distance matrix for the $CD$ cluster. 

<br>

The distance from $CD$ to $AB$,
$$
distance(CD, AB)=average(distance(CD, AB))=\frac{distance(AB, C) + distance(AB, D)}{2}=\frac{3+2}{2}=2.5
$$

<br>

The distance from $CD$ to $E$,
$$
distance(CD, E)=average(distance(CD, E))=\frac{distance(C, E) + distance(D, E)}{2}=\frac{3+5}{2}=4
$$

the new distance matrix with cluster $AB$:
$$
\begin{array}{|c|c|c|c|}
\hline
& AB & CD & E\\
\hline
AB & 0 & 2.5 & 3.5\\
\hline
CD &  & 0 & 4\\
\hline
E & & & 0\\
\hline
\end{array}
$$

<br>

Now we find the next minimum distance from table 3 to form the third cluster. The minimum distance from the table is 2.5 at $AB,CD$. Now we generate the distance matrix for the $ABCD$ cluster. 

<br>

The distance from $ABCD$ to $AB$,
$$
distance(ABCD, E)=average(distance(ABCD, E))=\frac{distance(AB, E) + distance(CD, E)}{2}=\frac{3.5+4}{2}=3.75
$$

<br>

the new distance matrix with cluster $AB$:
$$
\begin{array}{|c|c|c|}
\hline
& ABCD & E\\
\hline
ABCD & 0 & 3.75\\
\hline
E &  & 0\\
\hline
\end{array}
$$

<br>

The final merging distance from cluster $ABCD$ to $E$ is 3.75.
<br>

Final dendrogram:
<br>
```commandline
A   B   C   D   E       Distance
|   |   |   |   |           
 AB       CD    |           1
  |        |    |
     ABCD       |           2.5
       |        |
         ABCDE              3.75
```

### 1.3 Chapter 15

<br> 

#### Exercise 1: Consider Figure 15.12 and answer the following questions, assuming that we use the Euclidian distance between points, and that $\epsilon=2$ and $minpts=3$

<br>

#### a.) List all the core points

<br>

p, q, v, r, s, t, w, k, h, d, a, b, c, e, i, f, g, j, n, o

<br>

#### b.) Is $a$ directly density reachable from $d$?

<br>

Yes, because $a$ and $d$ are core points and lie within the given radius $\epsilon=2$

<br>

#### c.) Is $o$ density reachable from $i$? Show the intermediate points on the chain or the point where the chain breaks.

<br>

Yes, $o$ is density reachable from $i$. The intermediate points in the chain are: i, e, f, g, j, n, o

<br>

#### d.) Is density reachable a symmetric relationship, that is, if $x$ is reachable from $y$, does it imply that $y$ is density reachable from $x$? Why or why not?

<br>

No, density reachable is not a symmetric relationship. The reason for this is that if you have a core point x and y is density reachable from y through z, and if y is not a core point, then x is not density reachable from y. 

<br>

#### e.) Is $l$ density connected to $x$? Show the intermediate points that make them density connected or violate the property, respectively.

<br>

Yes, $l$ is density connected to $x$, the intermediate points that make them density connected is $w$, $t$. 

<br>

#### f.) Is density connected a symmetric relationship?

<br>

Yes. Density connected is a symmetric relationship.

<br>

#### g.) Shoe the density-based clusters and the noise points. 

<br>

There are two density based clusters.
<br>
Cluster 1: p, q, v, r, s, t, w, x, l, k, h, d, a
<br>
Cluster 2: m, i, e, b, c, f, g, j, n, o, u

<br>

#### Exercise 2: Consider the points in Figure 15.13. Define the following distance measures;
$$
L_{\infty}(x,y)=max^{d}_{i=1}{|x_i - y_i|}
$$

$$
L_{1/2}(x,y)=(\Sigma^{d}_{i=1} |x_i - y_i|^{1/2})^2
$$

$$
L_{min}(x,y)=min^{d}_{i=1}{|x_i - y_i|}
$$

$$
L_{pow}(x,y)=(\Sigma^{d}_{i=1} 2^{i-1} (x_i - y_i)^2)^{1/2}
$$

#### a.) Using $\epsilon=2$, $minpts=5$, and $L_{infty}$ distance, find all core, border, and noise points. 

<br>

Using $\epsilon=2$ and $minpts=5$, the following points are core points: b, c, e, f, and g. These points all have at least 5 points within their $\epsilon$-neighborhoods. The following points are border points: a, d, and h. These points are within the $\epsilon$-neighborhoods of core points, but they do not have at least five points within their own $\epsilon$-neighborhoods. The following points are noise points: i and j. These points are not within the $\epsilon$-neighborhoods of any core points.

<br>

#### b.) Show the shape of the ball radius $\epsilon=4$ using the $L_{1/2}$ distance. Using $minpts=3$ show all the clusters found by DBSCAN.

<br>

The shape of the ball of radius $\epsilon=4$ using the $L_{1/2}$ distance is a square with side length 8. This is because the $L_{1/2}$ distance is the same as the Euclidian distance when the dimensions of the data are equal. The points that fall within this square are a, b, c, e, f, g and d, h, i, j. Therefore, these points are all clustered together by DBSCAN. 

<br>

#### c.) Using $\epsilon=1$, $minpts=6$, and $L_{min}$, list all core, border, and noise points.

<br>

Using $\epsilon=1$ and $minpts=6$, the following points are core points: b, c, e, and f. These points all have at least 6 points within their $\epsilon$-neighborhoods. The following points are border points: a and g. These points are within the $\epsilon$-neighborhoods fo core points, but they do not have at least 6 points within their own $\epsilon$-neighborhoods. The following points are noise points: d, h, i, and j. These points are not within the $\epsilon$-neighborhoods of any core points. 

<br>

#### d.) Using $\epsilon=4$, $minpts=3$, and $L_{pow}$, show all clusters found by DBSCAN.

<br>

Using $\epsilon=4$ and $minpts=3$, the following clusters are found by DBSCAN: a,b, c, e, f, g, and d, h, i, j. This is because all of the points in each cluster have at least 3 points within their $\epsilon$-neighborhoods.

<br>

### 1.4 Chapter 17

<br>

#### Exercise 8: Show that the silhouette coefficients of a point lies in the interval [-1, +1].

<br>

The silhouette coefficient can be calculated by following these steps:
* let $a$ be the average dissimilarity of the data point with all other points in the same cluster
* let $b$ be the average dissimilarity of the data point with all points in the nearest neighboring cluster. 
* the silhouette coefficient for the data point is given by $\frac{b-a}{max(a,b)}$
<br>

To prove that the silhouette coefficient lies in the interval [-1,+1], we need to show that $(b-a)$ is bounded by $-max(a,b)$ and $+max(a,b)$.

* $max(a,b)$ is always greater than or equal to 0

<br>

$(b-a)$ has two cases:

* Case 1: $b \ge a$ 
  * In this case, $(b-a)$ is non-negative, or zero. The maximum value between $a$ and $b$ is $b$. Therefore, $\frac{b-a}{max(a,b)}$ is between 0 and 1.
* Case 2: $b < a$
  * In this case, $(b-a)$ is negative. The maximum value between $a$ and $b$ is $a$. Therefore, $\frac{b-a}{max(a,b)}$ is between -1 and 0.

<br>

From the above cases, we can observe that $\frac{b-a}{max(a,b)}$ is always between -1 and +1.

<br>

#### Exercise 10: Consider the dataset in Figure 17.10. Compute the silhouette coefficient for the point labeled $c$. Assume that the clusters are: $C_1={a,b,c,d,e}, C_2={g,i}, C_3={f,h,j}, C_4={k}$.

<br>

silhouette coefficient = $\frac{b-a}{max(a,b)}$

distance between points = $\sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}$

<br>

To find $a$:
<br>
$d(c,a)=2.9$
<br>
$d(e,d)=1.5$
<br>
$d(c,d)=2$
<br>
$d(c,e)=1$
<br>

$a=\frac{2.9+1.5+2+1}{4}=1.87$

<br>

for cluster 3:
<br>

$d(c,f)=4.2$
<br>
$d(c,h)=5$
<br>
$d(c,l)=4.9$
<br>
$b=\frac{4.2+5+4.9}{3}=4.6$
<br>

silhouette coefficient = $\frac{4.6-1.87}{4.6}=0.59$

<br>

for cluster 2:
<br>

$d(c,g)=5.8$
<br>
$d(c,i)=5$
<br>
$b=\frac{6+5}{2}=5.5$
<br>

silhouette coefficient = $\frac{5.5-1.85}{5.5}=0.65$


for cluster 4:
<br>

$d(c,k)=7$
<br>

silhouette coefficient = $\frac{7-1.87}{7}=0.73$

<br>

According to the coefficients, cluster 4 is the farthest and cluster 3 is the nearest.


## References

<br>

https://www.geeksforgeeks.org/k-means-clustering-introduction/

<br>

https://xlinux.nist.gov/dads/HTML/manhattanDistance.html#:~:text=Definition%3A%20The%20distance%20between%20two,y1%20%2D%20y2%7C.

<br>

https://www.cuemath.com/euclidean-distance-formula/

<br>

https://www.geeksforgeeks.org/ml-expectation-maximization-algorithm/

<br>

https://people.revoledu.com/kardi/tutorial/Similarity/SimpleMatching.html

<br>

https://www.statisticshowto.com/jaccard-index/

<br>

https://www.geeksforgeeks.org/dbscan-clustering-in-ml-density-based-clustering/

<br>

https://medium.com/@Suraj_Yadav/silhouette-coefficient-explained-with-a-practical-example-assessing-cluster-fit-c0bb3fdef719#:~:text=The%20silhouette%20coefficient%20is%20a,clusters

<br>


## Practicum Problems

### 2.1 Problem 1

#### Load the auto-mpg sample dataset from the UCI Machine Learning Repository into Python using a Pandas dataframe. Using only the continuous fields as features, impute any missing values with the mean, and perform a Hierarchical Clustering (Use sklearn.cluster.AgglomerativeClustering) with linkage set to average and the default affinity set to euclidean. Set the remaining parameters to obtain a shallow tree with 3 clusters as the target. Obtain the mean and variance values for each cluster, and compare these values to the values obtained for each class if we used origin as a class label. Is there a clear relationship between cluster assignment and class label?

In [1]:
import pandas as pd
from sklearn.cluster import AgglomerativeClustering
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

#load dataset
column_names = ['MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight', 'Acceleration', 'Model Year', 'Origin', 'Car Name']
data = pd.read_csv('auto-mpg.data', delim_whitespace=True, names=column_names, na_values='?')

#select only the continuous fields
continuous_columns = ['MPG', 'Displacement', 'Horsepower', 'Weight', 'Acceleration']
data_continuous = data[continuous_columns]

#impute missing values with the mean
mean_imputer = SimpleImputer(strategy='mean')
data_mean_imp = data_continuous.copy()
data_mean_imp['Horsepower'] = mean_imputer.fit_transform(data_mean_imp[['Horsepower']])

#scale the data
scaler = StandardScaler()
data_scaled = scaler.fit_transform(data_mean_imp)

#perform hierarchical clustering with 3 clusters
clustering = AgglomerativeClustering(n_clusters=3, linkage='average')
clusters = clustering.fit_predict(data_scaled)

#add cluster labels back to the dataframe
data['Cluster'] = clusters

#compute stats for each cluster
cluster_stats = data.groupby('Cluster')[['MPG', 'Displacement', 'Horsepower', 'Weight', 'Acceleration']].agg(['mean','var'])

#compute stats for each origin class
origin_stats = data.groupby('Origin')[['MPG', 'Displacement', 'Horsepower', 'Weight', 'Acceleration']].agg(['mean','var'])

#print results
print("Cluster stats:\n", cluster_stats)
print()
print("Origin stats:\n", origin_stats)

Cluster stats:
                MPG            Displacement               Horsepower   
              mean        var         mean          var        mean   
Cluster                                                               
0        26.177441  41.303375   144.304714  3511.485383   86.120275  \
1        14.528866   4.771033   348.020619  2089.499570  161.804124   
2        43.700000   0.300000    91.750000    12.250000   49.000000   

                          Weight                Acceleration            
                var         mean            var         mean       var  
Cluster                                                                 
0        294.554450  2598.414141  299118.709664    16.425589  4.875221  
1        674.075816  4143.969072  193847.051117    12.641237  3.189948  
2          4.000000  2133.750000   21672.916667    22.875000  2.309167  

Origin stats:
               MPG            Displacement               Horsepower   
             mean        var     

There seems to be a relationship between cluster assignments and origin class labels, especially for cluster 1 and possibly cluster 2. Cluster 1 aligns with origin 1 characteristics, where the cars typically have high displacement, high horsepower, heavier, and have lower fuel efficiency. Cluster 2 could possibly align with origin 3 in terms of high MPG and lower displacement. 

### 2.2 Problem 2

#### Load the Boston dataset (sklearn.datasets.load_boston()) into Python using a Pandas dataframe. Perform a K-means analysis on scaled data, with the number of clusters ranging from 2 to 6. Provide the Silhouette score to justify which value of k is optimal. Calculate the mean values for all features in each cluster for the optimal clustering - how do these values differ from the centroid coordinates?

In [2]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score


#since load_boston() no longer works on scikit learn, I instead loaded the csv file of the boston housing prices
#load dataset
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 
                'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
data = pd.read_csv('boston_house_prices.csv', header=None, names=column_names)

#scaling the data
scaler = StandardScaler()
scaled_data = scaler.fit_transform(data)

#perform k means analysis for cluster numbers from 2 to 6 and compute the silhouette score
inertia = []
cluster_range = range(2, 7)
silhouette_scores = []
for k in cluster_range:
    kmeans = KMeans(n_clusters=k, random_state=42)
    kmeans.fit(scaled_data)
    inertia.append(kmeans.inertia_)
    score = silhouette_score(scaled_data, kmeans.labels_)
    silhouette_scores.append(score)

print("Silhouette Scores:")
for i in range(0,5):
    print(silhouette_scores[i])
print()

#perform k means clustering with the optimal number of clusters, k=2
kmeans_optimal = KMeans(n_clusters=2, random_state=42)
kmeans_optimal.fit(scaled_data)

#assign the cluster labels to the dataframe
data['Cluster'] = kmeans_optimal.labels_

#calculate the mean values for all features in each cluster
cluster_means = data.groupby('Cluster').mean()

print(cluster_means)

Silhouette Scores:
0.35010701730151716
0.23388957997403387
0.24394269176812938
0.2455483784444133
0.22579996220030626

             CRIM         ZN      INDUS      CHAS       NOX        RM   
Cluster                                                                 
0        0.263946  17.477204   6.919818  0.069909  0.487215  6.456544  \
1        9.839575   0.000000  18.975085  0.067797  0.680124  5.965096   

               AGE       DIS        RAD         TAX    PTRATIO           B   
Cluster                                                                      
0        56.382067  4.751124   4.474164  302.209726  17.818237  386.643891  \
1        91.238418  2.017920  18.983051  605.316384  19.640113  300.967345   

             LSTAT       MEDV  
Cluster                        
0         9.417812  25.782067  
1        18.666610  16.493220  


Observing the silhouette scores, the most optimal number of clusters $k$ is 2, as it has the highest silhouette score of about 0.350. This suggests that clustering the dataset into two clusters results in the best separation and internal cohesion among the clusters compared to higher values of $k$.
<br>
These means values differ from the centroid coordinates. The centroids are calculated in the scaled space where each feature has been standardized (mean = 0, standard deviation = 1), thus the values of the centroids represent the position relative to the standard deviation of each feature across the entire dataset. The mean values are in the original scale of the data. The centroid values can be negative or positive reflecting their relative position from the mean (zero in scaled space). In contrast, the mean values reflect the actual average values of each feature within the clusters in the dataset's original units.


### 2.3 Problem 3

#### Load the wine dataset (sklearn.datasets.load_wine()) into Python using a Pandas dataframe. Perform a K-Means analysis on scaled data, with the number of clusters set to 3. Given the actual class labels, calculate the Homogeneity/Completeness for the optimal k - what information do each of these metrics provide?

In [59]:
import pandas as pd
from sklearn.datasets import load_wine
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn import metrics

#load dataset
wine = load_wine()
dataframe = pd.DataFrame(data=wine.data, columns=wine.feature_names)

#add target column to the dataframe
dataframe['target'] = wine.target

#scaling the features
scaler = StandardScaler()
scaled_features = scaler.fit_transform(dataframe)

#initialize K-Means with 3 clusters
kmeans = KMeans(n_clusters=3)
kmeans.fit(scaled_features)

#adding a cluster column to the dataframe to see what cluster a row is in
dataframe['cluster'] = kmeans.labels_

#assuming target from the dataset is the true label
true_labels = wine.target

#calculating homogeneity score and completeness score
homogeneity_score = metrics.homogeneity_score(true_labels, kmeans.labels_)
completeness_score = metrics.completeness_score(true_labels, kmeans.labels_)

print("Homogeneity score:", homogeneity_score)
print("Completeness score:", completeness_score)

Homogeneity score: 0.9126123393277744
Completeness score: 0.9091008506516705


A homogeneity score that is close to 1 means that the clusters are highly homogeneous. This means that most of the clusters consist almost of data points belonging to a single class. A completeness score close to 1 means that nearly all data points that are members of a specific class are grouped together into the same cluster. This indicates that the clustering model has successfully manged to include most members of each class within a single cluster. In our case, we have a high value for homogeneity and completeness. This means our clusters are pure and complete, there are no trade-offs where one metric is sacrificed to improve the other. 

## References

<br>

https://scikit-learn.org/stable/modules/clustering.html#hierarchical-clustering

<br>

https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/datasets/data/boston_house_prices.csv

<br>

https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html

<br>

https://www.geeksforgeeks.org/ml-handle-missing-data-with-simple-imputer/

<br>

https://stackoverflow.com/questions/51138686/how-to-use-silhouette-score-in-k-means-clustering-from-sklearn-library

<br>

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.homogeneity_score.html

<br>

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.completeness_score.html

<br>

ChatGPT